addons/udes_stock/tests/test_picking_batch.py

```diff
@@ -1032,6 +1032,157 @@ def test31_backorder_generated_unpickable_item_product_ok_in_two_picks_and_two_p
+            pickings, 2, "assigned", [4], package_list=apple_packages[3] + banana_packages[3]
+        )
+        # Check the state of the investigation pickings - should not be grouped
+        inv_pickings = Picking.search([("picking_type_id", "=", self.picking_type_internal.id)])
+        self._assert_apple_banana_pickings(
+            inv_pickings,
+            6,
+            "assigned",
+            [1, 2, 3],
+            package_list=apple_packages[:-1] + banana_packages[:-1],
+        )
+
+        # Check the state of the pick pickings
+        # Expect to have some draft pickings due to refactoring - which are deleted via cron
+        # Two confirmed pickings which have everything kicked out but grouped by product
+        # Two assigned pickings, which are the original ones
+        pick_pickings = Picking.search([("picking_type_id", "=", self.picking_type_pick.id)])
+        draft_pickings = pick_pickings.filtered(lambda p: p.state == "draft")
+        self.assertGreater(len(draft_pickings), 0)
+        unlink_cron.method_direct_trigger()
+        pick_pickings = Picking.search([("picking_type_id", "=", self.picking_type_pick.id)])
+        self.assertEqual(len(pick_pickings), 4)
+        confirmed_pickings = pick_pickings.filtered(lambda p: p.state == "confirmed")
+        self._assert_apple_banana_pickings(confirmed_pickings, 2, "confirmed", [1, 2, 3])
+        for p in confirmed_pickings:
+            with self.subTest(p=p):
+                self.assertTrue(p.backorder_id)
+                self.assertTrue(p.group_id)
+        remaining_pickings = pick_pickings - confirmed_pickings - draft_pickings
+        self.assertEqual(remaining_pickings, pickings)
+
+        # Flag the last item of each product as unpickable which should close the batch and
+        # result in a single picking for each product
+        for package in apple_packages[3] + banana_packages[3]:
+            batch.unpickable_item(
+                reason="damaged", package_name=package.name, raise_stock_investigation=True
+            )
+        self.assertEqual(batch.state, "done")
+        # Check the state of the investigation pickings - should not be grouped
+        inv_pickings = Picking.search([("picking_type_id", "=", self.picking_type_internal.id)])
+        self._assert_apple_banana_pickings(
+            inv_pickings, 8, "assigned", [1, 2, 3, 4], package_list=apple_packages + banana_packages
+        )
+        # Check the pickings are correct - two separate ones grouped by product, it merges
+        # everything into the created picking with the group.
+        # Trigger the cron job to clean up the draft pickings, and remove the backorder id
+        unlink_cron.method_direct_trigger()
+        pick_pickings = Picking.search([("picking_type_id", "=", self.picking_type_pick.id)])
+        self._assert_apple_banana_pickings(pick_pickings, 2, "confirmed", [1, 2, 3, 4])
+        for p in pick_pickings:
+            with self.subTest(p=p):
```

Comment:
Suggestion (feel free to resolve):

Might be worth refactoring the check for `backorder_id` and `group_id` into a method, as this is similar to the checks done on line 1156.

e.g. something like:

`confirmed_pickings.assert_backorder_and_group(expected_backorder=True, expected_group=True)`

and

`pick_pickings.assert_backorder_and_group(expected_backorder=False, expected_group=True)`

Reply:
Thanks, have left this as it was only used twice and having the branches to filter down the assertTrue/False route seemed more work than :)